# SQL Project

<b> Customer: </b>

- A company planning to purchase a large subscription book reading service.

<b> Data provided: </b>

   - Table "books" — contains data about books;
   - Table "authors" — contains data about authors;
   - Table "publishers" — contains data about publishers;
   - Table "ratings" — contains data on user ratings of books;
   - The "reviews" table — contains data about user reviews.

<b> Research objective: </b>
- Analysis of the provided databases in order to form a value proposition for a new product.

In [1]:
from IPython.display import Image
from IPython.core.display import HTML
print('__________________________________________SCHEME__________________________________________')
Image(url= "https://pictures.s3.yandex.net/resources/scheme_1589269096.png")

__________________________________________SCHEME__________________________________________


## Data Preparation and Exploration

In [2]:
import pandas as pd
from sqlalchemy import create_engine

In [3]:
# set parameters
db_config = {'user': 'praktikum_student', # user name
 'pwd': 'Sdf4$2;d-d30pp', # password
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # connection port
 'db': 'data-analyst-final-project-db'} # name of the database
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# saving the connector
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [4]:
# Create a function to execute queries
def request(query):
    return pd.io.sql.read_sql(query, con = engine) 

In [5]:
# List of tables
table_list = ['books', 'authors', 'publishers', 'ratings', 'reviews']

In [6]:
# In the loop, let's run through the list of tables, display the first five rows and information about each table
for table in table_list:
    print('Таблица "{}"'.format(table))
    display(request('SELECT * FROM {}'.format(table)).head())
    display(request('SELECT * FROM {}'.format(table)).info())
    print('-----------------------------------------------------------------------------------------------------')
    print()

Таблица "books"


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

-----------------------------------------------------------------------------------------------------

Таблица "authors"


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

-----------------------------------------------------------------------------------------------------

Таблица "publishers"


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

-----------------------------------------------------------------------------------------------------

Таблица "ratings"


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

-----------------------------------------------------------------------------------------------------

Таблица "reviews"


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

-----------------------------------------------------------------------------------------------------



### Summary
Tables with data have been viewed, information about them has been studied. You can start completing tasks.

## Solving tasks

### Number of books published after January 1, 2000

In [7]:
request(
'''
SELECT 
    COUNT(book_id) AS books_published_after_1_january_2000
FROM 
    books 
WHERE 
    CAST(publication_date AS date) > '2000-01-01'
'''
)

books_published_after_1_january_2000
0                                   819

#### Summary
819 books were released after January 1, 2000.

### Number of reviews and average rating of books

In [8]:
request(
'''
SELECT
    books.title AS title,
    subq.count AS cnt_reviews,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
JOIN 
    ratings ON books.book_id = ratings.book_id
JOIN (
    SELECT
        books.book_id,
        books.title,
    COUNT
        (reviews.review_id)
    FROM 
        books
    LEFT JOIN 
        reviews ON books.book_id = reviews.book_id
    GROUP BY 
        books.book_id) AS subq ON books.book_id = subq.book_id
    GROUP BY 
        books.book_id, subq.count
    ORDER BY 
        avg_rating DESC
'''
)

title  cnt_reviews  \
0    The Adventures of Tom Sawyer and Adventures of...            1   
1                      Act of Treason (Mitch Rapp  #9)            2   
2              School's Out—Forever (Maximum Ride  #2)            3   
3                 Stolen (Women of the Otherworld  #2)            2   
4                 Piercing the Darkness (Darkness  #2)            2   
..                                                 ...          ...   
995  The World Is Flat: A Brief History of the Twen...            3   
996                                      Drowning Ruth            3   
997                                              Junky            2   
998                  His Excellency: George Washington            2   
999                               Harvesting the Heart            2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

#### Summary
The number of reviews and the average score for each book are calculated. I wonder why I didn't like the book "Harvesting the Heart" so much?

### The publisher that has published the largest number of books thicker than 50 pages (exclusion from the analysis of brochures)

In [9]:
request(
'''
SELECT 
    publishers.publisher AS publishing_house,
    COUNT(books.book_id) AS number_of_books
FROM 
    publishers 
    JOIN books ON books.publisher_id = publishers.publisher_id
WHERE 
    books.num_pages > 50
GROUP BY 
    publishers.publisher_id
ORDER BY 
    COUNT(books.book_id) DESC
'''
)

publishing_house  number_of_books
0                   Penguin Books               42
1                         Vintage               31
2        Grand Central Publishing               25
3                Penguin Classics               24
4                Ballantine Books               19
..                            ...              ...
329         Ballantine Books (NY)                1
330                 Plaza y Janés                1
331       HarperCollinsPublishers                1
332  Random House: Modern Library                1
333                 Pan Childrens                1

[334 rows x 2 columns]

#### Summary
Penguin Books has published 42 books (with a volume of more than 50 pages) and this is the highest result among all the publishers represented.

### The author with the highest average rating of books (books with 50 or more ratings)

In [10]:
request(
'''
SELECT 
    authors.author,
    AVG(ratings.rating) AS avg_rating
FROM 
    ratings
    JOIN books ON books.book_id = ratings.book_id
    JOIN authors ON books.author_id = authors.author_id
WHERE 
    ratings.book_id IN (
    SELECT
        ratings.book_id 
    FROM 
        ratings
    GROUP BY 
        ratings.book_id
    HAVING 
        COUNT(ratings.book_id) >= 50
)
GROUP BY 
    authors.author_id
ORDER BY 
    AVG(ratings.rating) DESC
'''
)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.287097
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.246914
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                          Lois Lowry    3.750000
10                                          Dan Brown    3.741259
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

#### Summary
The author with the highest average rating of books is J.K. Rowling (with illustrations by Mary GrandPré). The rating of the famous author is 4.3

### Average number of reviews from users who put more than 50 ratings

In [11]:
request(
'''
SELECT
    AVG(SUBQ2.review_count) AS avg_review
FROM (
    SELECT 
        reviews.username,
        COUNT(reviews.review_id) AS review_count
    FROM 
        reviews
    RIGHT JOIN (
        SELECT 
            ratings.username,
            COUNT(ratings.rating_id) AS rating_count
        FROM 
            ratings
        GROUP BY 
            ratings.username
        HAVING 
            COUNT(ratings.rating_id) >= 50) AS SUBQ1 ON SUBQ1.username = reviews.username
    GROUP BY reviews.username) AS SUBQ2
'''
)

avg_review
0   24.222222

#### Summary
On average, 24 reviews were prepared by users who gave more than 50 ratings.

## General Conclusions and Recommendations

The conducted research of the data contained in the service has established:
- The service contains 1000 books:
- most of them - 819 books were released after 01.01.2000;
 - Books written by 636 authors: 
     - the most highly appreciated author by users - J.K. Rowling (4.3 points);
 - Collected books were produced by 340 publishing houses: 
     - Penguin Books has published 42 books (with a volume of more than 50 pages) and this is the highest result among all the publishers represented;
 - 6456 ratings of books put down by users: 
     - the lowest score for the book Harvesting the Heart (1.5 points);
 - 2793 reviews prepared by users 
     - 24 reviews on average were prepared by users who gave more than 50 ratings.